### DISCUSSION TITLE
What is needed to call `createModelReference`

### DISCUSSION THREAD

**Discussion by alissa-tung** [CodinGame/monaco-vscode-api#403] [Q&A] [answered]
Hello, I am trying to use this library to develop a web editor. Thank you for developing this.

My question is, what is needed and to be prepared to call `createModelReference`?

I tried to call `createModelReference` with a `file://` scheme url, it says `Model not found`, the stack trace shows that was by `modelService.getModel`

I tried to call `createModelReference` with a `inmemory://` scheme url, it says `ENOPRO: No file system provider found for resource`

**Comment by CGNonofr** [comment, marked as answer]
createModelReference should only be called on a file that already exists on the virtual filesystem, so it won't work for your `inmemory` model indeed, you have 2 solutions:
- the function takes a second argument with the content of the file, it will be written on the memory filesystem just before actually calling createModelReference, but I don't recommend it depending on your needs (before it's hard to clean the file that will remain in the memory filesystem otherwise forever)
- register your files [just like in the demo](https://github.com/CodinGame/monaco-vscode-api/blob/2819ccd49ce36f9515e618b5d451a1972b1eb1e0/demo/src/setup.common.ts#L69C7-L150). it allows you to remove your filesystem overlay when you don't need it anymore. You can use the default `RegisteredFileSystemProvider` or use any other implementation of yours if you want

**Reply by alissa-tung** [reply]
Thank you! I think the two solutions clear my questions on the in memory scheme files.

I am wondering how can I have a Monaco editor instance, that can handle all `file://` scheme models, which is maps to the real disk file system on someone's machine. Do we have document or service overrides on this?

**Reply by CGNonofr** [reply]
There is 3 ways depending on what you are exactly trying to achieve
- if you want to use the filesystem of the machine the browser runs on, you can use the HTMLFileSystemProvider (what is used by vscode.dev), [the demo here show how to use it](https://github.com/CodinGame/monaco-vscode-api/blob/2819ccd49ce36f9515e618b5d451a1972b1eb1e0/demo/src/main.common.ts#L99-L109) but it has some constraints:
  - it doesn't work in Firefox
  - You can only create it during a user interaction (click, key press...) on the browser will probably block it
- You implement by a custom protocol a way to access the filesystem by any mean you want, and implement a `IFileSystemProviderWithFileReadWriteCapability` by hands
- You use the vscode extension host server which allow to access a remote machine filesystem (among other things). Everything's explained [in the doc](https://github.com/CodinGame/monaco-vscode-api/blob/main/docs/vscode_server.md)

**Reply by alissa-tung** [reply]
Thank you a lot, I will try to implement a `IFileSystemProviderWithFileReadWriteCapability`, then `registeredFileSystemProvider` for it.

**Reply by alissa-tung** [reply]
BTW, I am trying to build an web editor using Next.js server actions, which are functions run on nodejs runtime, I am going to search if there is implementation of `IFileSystemProviderWithFileReadWriteCapability` on nodejs. Thank you for explaining!

**Comment by alissa-tung** [comment]
Hello @CGNonofr , I am trying to write a application that maps `file://` url to that one on disk file system. I still have a few questions, could you help me fix it? thank you

I had implement the `IFileSystemProviderWithFileReadWriteCapability`, and `registerFileSystemOverlay(1, fileSystemProvider)`. But when I call `createModelReference`, it says that `Error: Model not found`. I am wondering did I miss something, like,

- should I use `RegisteredFileSystemProvider` to get the custom provider that I had just registered, then register every file before `createModelReference`?
- I noticed that the readme says that we had include files service by default. I am thinking about what does this means to be default included? does this refers to that I do not needs to `getServiceOverride` for files service?

**Reply by CGNonofr** [reply]
> should I use RegisteredFileSystemProvider to get the custom provider that I had just registered, then register every file before createModelReference?

Not sure what you mean, but you don't need RegisteredFileSystemProvider if you implement IFileSystemProviderWithFileReadWriteCapability directly

> I noticed that the readme says that we had include files service by default. I am thinking about what does this means to be default included? does this refers to that I do not needs to getServiceOverride for files service?

It just means you don't need to call getServiceOverride and provide the result to the initialize function, the initialize function already does it

Did you register the `model` service override though?

**Reply by alissa-tung** [reply]
Thank you, I finally get managed to write an React server actions implement of `IFileSystemProviderWithFileReadWriteCapability`.

You explanation perfectly solved the problems that I encountered.

- I was confusing about the use of `RegisteredFileSystemProvider`. And from you answer now I am sure that `RegisteredFileSystemProvider` is not needed to be called if I provide a custom one.
- My problem is that I have not register the `model` service override. Thank you for reminding on this. It is needed to register it.

I think I am fully get well with the `createModelReference` . The hack parts was by React. Thank you!
